tests/test_datasets/test_utils.py

```diff
@@ -514,3 +517,298 @@ def test_timestamp_range_fail_type(start, end, periods, freq):
+        ),
+        ("df_aligned_int", {"segment_0": 19, "segment_1": 19}, None, {"timestamp", "segment", "target"}),
+        (
+            "df_aligned_int",
+            {"segment_0": 19, "segment_1": 19},
+            "original_timestamp",
+            {"timestamp", "segment", "target", "original_timestamp"},
+        ),
+    ],
+)
+def test_apply_alignment_format(df_name, alignment, original_timestamp_name, expected_columns, request):
+    df = request.getfixturevalue(df_name)
+    result_df = apply_alignment(df=df, alignment=alignment, original_timestamp_name=original_timestamp_name)
+
+    assert len(result_df) == len(df)
+    assert set(result_df.columns) == expected_columns
+
+
+@pytest.mark.parametrize(
+    "df_name, alignment",
+    [
+        (
+            "df_aligned_datetime",
+            {"segment_0": pd.Timestamp("2020-01-10"), "segment_1": pd.Timestamp("2020-01-10")},
+        ),
+        (
+            "df_aligned_datetime",
+            {"segment_0": pd.Timestamp("2020-01-10"), "segment_1": pd.Timestamp("2020-01-05")},
+        ),
+        (
+            "df_aligned_datetime",
+            {"segment_0": pd.Timestamp("2020-01-05"), "segment_1": pd.Timestamp("2020-01-10")},
+        ),
+        ("df_aligned_int", {"segment_0": 19, "segment_1": 19}),
+        ("df_aligned_int", {"segment_0": 19, "segment_1": 14}),
+        ("df_aligned_int", {"segment_0": 14, "segment_1": 19}),
+    ],
+)
+def test_apply_alignment_doesnt_change_original(df_name, alignment, request):
+    df = request.getfixturevalue(df_name)
+    result_df = apply_alignment(df=df[::-1], alignment=alignment, original_timestamp_name="original_timestamp")
+
+    check_df = result_df.drop(columns=["timestamp"])
+    check_df = check_df.rename(columns={"original_timestamp": "timestamp"})
+    pd.testing.assert_frame_equal(check_df.loc[df.index], df)
+
+
+@pytest.mark.parametrize(
+    "df_name, alignment, expected_timestamps",
+    [
+        (
+            "df_aligned_datetime",
+            {"segment_0": pd.Timestamp("2020-01-10"), "segment_1": pd.Timestamp("2020-01-10")},
+            list(range(-9, 1)) + list(range(-9, 1)),
+        ),
+        (
+            "df_aligned_datetime",
+            {"segment_0": pd.Timestamp("2020-01-10"), "segment_1": pd.Timestamp("2020-01-05")},
+            list(range(-9, 1)) + list(range(-4, 6)),
+        ),
+        (
+            "df_aligned_datetime",
+            {"segment_0": pd.Timestamp("2020-01-05"), "segment_1": pd.Timestamp("2020-01-10")},
+            list(range(-4, 6)) + list(range(-9, 1)),
+        ),
+        ("df_aligned_int", {"segment_0": 19, "segment_1": 19}, list(range(-9, 1)) + list(range(-9, 1))),
+        ("df_aligned_int", {"segment_0": 19, "segment_1": 14}, list(range(-9, 1)) + list(range(-4, 6))),
+        ("df_aligned_int", {"segment_0": 14, "segment_1": 19}, list(range(-4, 6)) + list(range(-9, 1))),
+    ],
+)
+def test_apply_alignment_new_timestamps(df_name, alignment, expected_timestamps, request):
```

Comment:
Should we check `df_misaligned_datetime` and `df_misaligned_into` too?

Reply:
Ok, I'll add one couple test cases for them just in case.